Refactor baselines: add generic trainer and common utils

JasonJiazhiZhang opened the pull request:
## Looking for Suggestions
Looking for a good name that competently describes [this config node](https://github.com/JasonJiazhiZhang/habitat-api/blob/0a7ff0280dbeb7e4a4f108c83f4f0e9b7808e446/habitat_baselines/config/default.py#L24) and all child nodes: used to be `BASELINE` and now renamed to `TRAINER`. Alternatives: `LEARNER`, `EXP(ERIMENT)`, `MODEL(S)`. 
Suggestions are greatly appreciated!
## Motivation and Context

This PR aims to bring modularity and improve reusability of habitat_baseline,  by introducing abstraction to the current experiment process and collecting common utilities. This is an early draft and all suggestions are welcomed :)
Since the amount of code changed is relatively large, here's a breakdown of what does this PR do, what directions am I particularly looking for suggestions/opinions, and what needs to be done before merging.

* **What does this PR do?**
  1. Provide a trainer template (`BaseTrainer`) and  create trainer registry.  I went through a couple of related repos ([Horizon](https://github.com/facebookresearch/Horizon), [stable_baselines](https://github.com/hill-a/stable-baselines), [Pythia](https://github.com/facebookresearch/pythia)) and chose a high-level structure similar to [this](https://github.com/hill-a/stable-baselines/blob/master/stable_baselines/common/base_class.py) in stable_baselines (suggestions on implementation details are welcomed! ). This will  make it more straightforward to add new trainers. Note that this is complementory to #144 , as that one is more specific to ppo and aims to make it extensible to new tasks. 
  2. Collect trainer-independent utilities, such as env creation, batch transpose, and tensorboard logging, and move them to `habitat_baselines/common/` to give better reusability. 
  3. Merge logics in train_ppo and evaluate_ppo into one `ppo_trainer.py` file. Now launch commands for training/evaluation look like this:
  ```bash
  python  habitat_baselines/run.py --exp-config /path/to/config/ppo.yaml --run-type train
  ```
  4. Unify env_config, baseline_config, and previous command-line args into one config system so that it can be passed and stored in a single place.
  Here's a directory tree that might help visualize the structural changes:
```diff
habitat_baselines/
 ├── __init__.py
+├── common/                      	# directory that includes reusable functionalities
+│   ├── base_trainer.py           	# trainer template
+│   ├── env_utils.py
+│   ├── environments.py                # customized RL environments
+│   ├── tensorboard_utils.py
+│   ├── baseline_registry.py           # registry for envs and trainers
+│   └── utils.py
 ├── config/
+│   └── pointnav/
+│       └── ppo.yaml 
 ├── rl/
 │   ├── __init__.py
 │   ├── ppo/
 │   │   ├── __init__.py
 │   │   ├── policy.py
 │   │   ├── ppo.py
+│   │   ├── ppo_trainer.py  		# trainer class for ppo (train and eval)
-│   │   └── utils.py
 │   └── requirements.txt
+├── run.py                             # main entry point to all experiments
-├── evaluate_ppo.py
-├── train_ppo.py
 └── ...
```

* **Things need to be delivered before merging:**
  - [x] Polishing, docs, typing, header  
  - [x] Fix imports in habitat_baselines `__init__` 
  - [ ] change README
* **TODO (maybe in future PRs):**
     - Use saved configs in ppo_trainer eval(), and make compatible with old checkpoints.
     - Merge with #144.
     - Reorganize the config system :|

## How Has This Been Tested
Current tests pass and train/eval run properly from `run.py`
<!--- Please describe here how your modifications have been tested. -->

## Types of changes

<!--- What types of changes does your code introduce? Leave all the items that apply: -->

- Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

JasonJiazhiZhang commented:
For now this PR is expected to fail CI test.
This command below in  circle-ci `run api tests` will have to be modified accordingly for it to pass. 
```
python -u habitat_baselines/train_ppo.py  ...
```
Maybe change to something like this when PR lands:
```
python -u habitat_baselines/run.py --exp-config habitat_baselines/config/pointnav/ppo.yaml --run-type train
```

abhiskk commented on habitat_baselines/common/base_trainer.py:
```diff
@@ -0,0 +1,48 @@
+from habitat_baselines.common.trainer_registry import trainer_registry
+
+
+class BaseTrainer(ABC):
```
To stay uniform with other places in the codebase, let's avoid inheriting from `ABC` and instead using, `raise NotImplementedError`. 

Note: The ABC design is definitely valid but it is something which the libraries from our domain don't use (eg: pytorch), hence `ABC` was not included in the original design.

abhiskk commented on habitat_baselines/common/tensorboard_utils.py:
```diff
@@ -1,8 +1,30 @@
-from torch.utils.tensorboard import SummaryWriter
+
+
+class DummyWriter:
```
We can use `SummaryWriter = None` instead of using a `DummyWriter` here and handle the logic later.

JasonJiazhiZhang replied:
I think I'll create a separate PR for this. This PR did not change any of the tensorboard-specific logic and inherited directly from #127. I do agree that DummyWriter is a bit clunky and if we properly check `None` when appropriate we can get rid of the wrapper altogether.

abhiskk replied:
Can you create a TODO for above.

abhiskk commented on habitat_baselines/common/env_utils.py:
```diff
@@ -0,0 +1,141 @@
+from habitat import Config, SimulatorActions, make_dataset
+
+
+class NavRLEnv(habitat.RLEnv):
```
Place `NavRLEnv` inside a separate file under `common`.

mathfac replied:
Looks like we will need Registry for RLEnvs and call this PointGoalRLEnv. Also add meaningful comment how RLEnvs should be used.

apsdehal replied:
Ideally the registry should support nesting, something like `registry.register_env('rl.point_nav', EnvClassHere)`.

abhiskk commented on habitat_baselines/common/base_trainer.py:
```diff
@@ -0,0 +1,48 @@
+from habitat_baselines.common.trainer_registry import trainer_registry
+
+
+class BaseTrainer(ABC):
```
We should think about more useful abstract methods to add to the `BaseTrainer`, @JasonJiazhiZhang if possible can you communicate on the PR what are the implementations you referred to when building out the base trainer and what you eventually ended up using.

JasonJiazhiZhang replied:
Sure, this draft PR is precisely for discussions like this. This BaseTrainer structure is inspired mainly by the base_class.py in [stable-baselines](https://github.com/hill-a/stable-baselines/blob/master/stable_baselines/common/base_class.py). I will put all the sources I looked in the description.

mathfac commented on habitat_baselines/__init__.py:
```diff
@@ -3,3 +3,10 @@
+# import for proper registry
+from habitat_baselines.rl.ppo.ppo_trainer import PPOTrainer, RolloutStorage
+
+__all__ = ["PPOTrainer", "RolloutStorage"]
```
Include general Trainer as well to make baseline extension easier.

mathfac commented on habitat_baselines/common/env_utils.py:
```diff
@@ -0,0 +1,141 @@
+        return self.habitat_env.get_metrics()
+
+
+def make_env_fn(config, env_class, rank):
```
Duplication of code from utils.py

mathfac commented on habitat_baselines/common/utils.py:
```diff
@@ -0,0 +1,148 @@
+from habitat_baselines.config.default import get_config as baseline_cfg
+
+
+class Flatten(nn.Module):
```
All this classes need docs, when it should be used.

JasonJiazhiZhang replied:
Already on TODO list :)

mathfac commented on habitat_baselines/common/utils.py:
```diff
@@ -0,0 +1,148 @@
+
+
+def get_exp_config(cfg_path, opts=None):
+    config = Config(new_allowed=True)
```
How `new_allowed`  works?

JasonJiazhiZhang replied:
`new_allowed=True` allows merging configs even if they have non-overlapping config nodes. For example, merging baseline_config into env_config with `new_allowed=False` won't work since env_config by default doesn't have the `BASELINE` node.

mathfac replied:
@JasonJiazhiZhang is that new yacs feature? It wasn't there before.

JasonJiazhiZhang replied:
@mathfac I don't think it's new. See [here](https://github.com/rbgirshick/yacs/blob/aabe5bd54d193996ceedcff650cff10c38b1f90b/yacs/config.py#L74).

mathfac commented on habitat_baselines/common/utils.py:
```diff
@@ -0,0 +1,148 @@
+    return env
+
+
+def batch_obs(observations):
```
Maybe outside of this PR, this can be refactored to be more general function. Maybe we can re-use t[he structured array functionality from Numpy ](https://jakevdp.github.io/PythonDataScienceHandbook/02.09-structured-data-numpy.html#RecordArrays:-Structured-Arrays-with-a-Twist)or find analogy in Pytorch. Conversion of array of structures to structure of arrays is common pattern in Data science.

mathfac commented on habitat_baselines/config/default.py:
```diff
@@ -22,13 +22,51 @@
 # BASELINE
 # -----------------------------------------------------------------------------
 _C.BASELINE = CN()
+_C.BASELINE.TRAINER_NAME = "ppo"
```
```suggestion
_C.TRAINER.TYPE = "ppo"
```
Let's be consistent with habitat-api naming: https://github.com/facebookresearch/habitat-api/blob/master/habitat/config/default.py#L61. Maybe, even `BASELINE` -> `TRAINER`.

JasonJiazhiZhang replied:
> Maybe, even BASELINE -> TRAINER

This is definitely something I would like to discuss. I'm thinking something like `BASELINE` -> `EXPERIMENT`.

mathfac commented on habitat_baselines/rl/ppo/ppo_trainer.py:
```diff
@@ -0,0 +1,738 @@
+        )
+
+
+def generate_video(
```
Maybe, move to ppo_uitils or etc all other classes/functions except PPOTrainer.  PPOTrainer already large enough.

mathfac commented on habitat_baselines/run.py:
```diff
@@ -0,0 +1,18 @@
+from habitat_baselines.common.utils import experiment_args, get_exp_config
+
+
+def main():
```
I really like simplicity of the entry point and how easy we can contribute more baselines without messing other ones.

mathfac commented:
For CI, can you rebase on latest habitat-api changes? Thanks

mathfac commented:
Master branch CI is passing: https://circleci.com/gh/facebookresearch/habitat-api.

apsdehal commented on habitat_baselines/common/base_trainer.py:
```diff
@@ -0,0 +1,48 @@
+    return trainer(trainer_cfg)
+
+
+class BaseRLTrainer(BaseTrainer):
```
Is there a point in having this when it is not adding any new functionality on top of `BaseTrainer`?

JasonJiazhiZhang replied:
@apsdehal The rationale is that `common/base_trainer.py` should in the future support imitation learning and SLAM based agents and `BaseTrainer` would be a template for those. I do think we need more discussion on what functionalities to be included inside `BaseTrainer` vs. `BaseRLTrainer`

apsdehal commented on habitat_baselines/common/env_utils.py:
```diff
@@ -0,0 +1,141 @@
+
+
+class NavRLEnv(habitat.RLEnv):
+    def __init__(self, config_env, config_baseline, dataset):
```
Comments here would be helpful to explain more about the parameters.

apsdehal replied:
Also, it is unnatural to write it as `config_env` and `config_baseline` instead of `env_config` and `baseline_config`.

apsdehal replied:
And instead of `baseline_config`, it should be `model_config` now.

apsdehal replied:
Also, why this isn't single config instead of being `env_config` separately and `baseline_config` separately.

JasonJiazhiZhang replied:
@apsdehal This naming choices are largely inherited from the original [`train_ppo.py`](https://github.com/facebookresearch/habitat-api/blob/782f1f9aec09895dc27347927f01998ba383b8b4/habitat_baselines/train_ppo.py#L25). I do have some alternatives in mind (see [comments here](https://github.com/facebookresearch/habitat-api/pull/153#discussion_r302721173)). Also, as to why it's not a single config yet: I definitely intend to merge two configs (see [here](https://github.com/JasonJiazhiZhang/habitat-api/blob/f6478167b3e46b897e4b7b1158b265b63a3b6c31/habitat_baselines/common/env_utils.py#L76) and [here](https://github.com/JasonJiazhiZhang/habitat-api/blob/f6478167b3e46b897e4b7b1158b265b63a3b6c31/habitat_baselines/common/utils.py#L90-L97), comparing to [this](https://github.com/JasonJiazhiZhang/habitat-api/blob/782f1f9aec09895dc27347927f01998ba383b8b4/habitat_baselines/train_ppo.py#L92)), and should modify NavRLEnv for use one config for sure. Thank you for catching this 👍 .  I'll make those changes to NavRLEnv to  take one unified config, if people agree with this change.

apsdehal commented on habitat_baselines/common/env_utils.py:
```diff
@@ -0,0 +1,141 @@
+        super().__init__(config_env, dataset)
+
+    def reset(self):
+        self._previous_action = None
```
Does `self._episode_distance_covered` also need to be set to `None`. It probably won't affect anything but just in case.

JasonJiazhiZhang replied:
I think setting to None makes sense. @abhiskk @mathfac did you guys choose to not reset it on purpose?

apsdehal commented on habitat_baselines/common/env_utils.py:
```diff
@@ -0,0 +1,141 @@
+
+    def get_reward_range(self):
+        return (
+            self._config_baseline.BASELINE.RL.SLACK_REWARD - 1.0,
```
Same here. These should be `MODEL` instead of `BASELINE`. I might be missing some context though.

apsdehal replied:
So, if I have to use my own model, I will have to create a config with `BASELINE` attribute in it?

JasonJiazhiZhang replied:
If someone implements another algorithm, say A2C, they will create algorithm-specific config under `BASELINE.RL.A2C`. In this particular case tho, `SLACK_REWARD` is not algorithm-specific and is more related to the task and environment, so it lives under `BASELINE.RL`. Then again, it's probably a good idea to have a more informative/representative name than **BASELINE**

apsdehal commented on habitat_baselines/common/env_utils.py:
```diff
@@ -0,0 +1,141 @@
+        )
+
+    def get_reward(self, observations):
+        reward = self._config_baseline.BASELINE.RL.SLACK_REWARD
```
Comments.

apsdehal commented on habitat_baselines/common/env_utils.py:
```diff
@@ -0,0 +1,141 @@
+
+def construct_envs(config: Config, env_class):
+
+    random.seed(config.SEED)
```
Seed only in one single place. This can create mess later.

apsdehal commented on habitat_baselines/common/env_utils.py:
```diff
@@ -0,0 +1,141 @@
+        return self.habitat_env.get_metrics()
+
+
+def make_env_fn(config, env_class, rank):
```
Some general code philosophy here: If you are using function such as `make_*` be consistent with that. In trainer above, you added `get_trainer`, either change it to `make_trainer` or change everything else to `get_*`

JasonJiazhiZhang replied:
`make_env_fn` will be removed in a later PR as datasets will support scene split. Will stick to `get_*`

apsdehal commented on habitat_baselines/common/base_trainer.py:
```diff
@@ -0,0 +1,48 @@
+        pass
+
+
+def get_trainer(trainer_name, trainer_cfg):
```
At this point, i will suggest to move all of the `make_*` and `get_*` functions to a separate file in utils. They are all over the place now.

apsdehal commented on habitat_baselines/common/env_utils.py:
```diff
@@ -0,0 +1,141 @@
+        )
+
+    scene_splits = [[] for _ in range(baseline_cfg.num_processes)]
+    for j, s in enumerate(scenes):
```
```suggestion
    for idx, scene in enumerate(scenes):
```

apsdehal commented on habitat_baselines/common/env_utils.py:
```diff
@@ -0,0 +1,141 @@
+
+        agent_sensors = baseline_cfg.sensors.strip().split(",")
+        for sensor in agent_sensors:
+            assert sensor in ["RGB_SENSOR", "DEPTH_SENSOR"]
```
Will you keep updating this list in future if something changes. This ideally should be somewhere in the default config.

apsdehal commented on habitat_baselines/common/trainer_registry.py:
```diff
@@ -0,0 +1,27 @@
+from habitat.core.registry import Registry
+
+
+class TrainRegistry(Registry):
```
Why create a new registry class? Whole point of Registry pattern is to create a singleton which is easy to maintain instead of multiple sources of global truth. I would suggest using better functionalities in Registry itself to encapsulate everything there. For reference, check Pythia's registry again.

apsdehal commented on habitat_baselines/common/utils.py:
```diff
@@ -0,0 +1,148 @@
+        return CustomFixedCategorical(logits=x)
+
+
+def _flatten_helper(t, n, tensor):
```
Wasn't this define above also? And why do we have `Flatten` as well as `_flatten_helper`. A single class should be enough to handle both of the usages.

JasonJiazhiZhang replied:
Looks like another design choice inherited from old code that I did not pay much attention when refactoring. @abhiskk and @mathfac maybe this is something I can clean up?

erikwijmans replied:
The `_flatten_helper` is a helper function for creating the batches to learn from the rollout.  It's very specific to that class and probably belongs as a static method on the class.

apsdehal commented on habitat_baselines/common/utils.py:
```diff
@@ -0,0 +1,148 @@
+    return tensor.view(t * n, *tensor.size()[2:])
+
+
+def update_linear_schedule(optimizer, epoch, total_num_epochs, initial_lr):
```
Change this to a class `LRScheduler.

abhiskk replied:
@JasonJiazhiZhang  please add a TODO around this, it is a good suggestion.

apsdehal commented on habitat_baselines/common/utils.py:
```diff
@@ -0,0 +1,148 @@
+
+def make_env_fn(config, env_class, rank):
+    dataset = make_dataset(config.DATASET.TYPE, config=config.DATASET)
+    config.defrost()
```
This is a comment on older habitat design. Seeing defrost and freeze being repeatedly and exhaustively everywhere has defeated the whole purpose of `freeze`. It is time to reconsider that.

erikwijmans commented on habitat_baselines/rl/ppo/ppo_trainer.py:
```diff
@@ -0,0 +1,713 @@
+        )
+
+
+class RolloutStorage:
```
The RolloutStorage isn't specific to PPO and should be in a general utils.

JasonJiazhiZhang replied:
@erikwijmans I'll put it under `habitat_baselines/common/utils.py` for now and make `_flatten_helper` a static class method. Thanks!

mathfac commented:
CI test `python -u habitat_baselines/run.py --exp-config habitat_baselines/config/pointnav/ppo_train_test.yaml --run-type train` is trying to load checkpoint:
`FileNotFoundError: [Errno 2] No such file or directory: 'data/checkpoints/ckpt.0.pth'`

JasonJiazhiZhang commented:
> `FileNotFoundError: [Errno 2] No such file or directory: 'data/checkpoints/ckpt.0.pth'`
Looks like I didn't check for dir existence, thanks for spotting that.

erikwijmans commented on habitat_baselines/config/pointnav/ppo.yaml:
```diff
@@ -0,0 +1,40 @@
+      use_linear_clip_decay: True
+      use_linear_lr_decay: True
+      reward_window_size: 50
+      sensors: "RGB_SENSOR,DEPTH_SENSOR"
```
This should be a list

JasonJiazhiZhang replied:
I think this not being a list originally is to support the command line option augmentation [feature](https://github.com/facebookresearch/habitat-api/blob/732377878c8c4b2503358e184bac4b17f2a8937e/habitat/config/default.py#L188). I kept the feature to add config option via command line so maybe a string is better suited here.

erikwijmans replied:
The `video_option` field is a list tho?  Is that not able to be changed and should be a comma separated string?

JasonJiazhiZhang replied:
I think that should also be a string instead for consistency. Will fix :)

mathfac replied:
Agree with Erik it should be list, but is in scope of next RL refactoring PR.

apsdehal replied:
Command line arguments from yacs support list since it does literal eval on what you pass. So this can be converted to an array.

JasonJiazhiZhang replied:
@apsdehal Thanks for pointing out! I didn't test `option '["a","b"]'` and turns out that works for literal eval. Updated [here](https://github.com/facebookresearch/habitat-api/pull/172/commits/bba7ac81591d447eb31be2623e6c69b9fcec6840) in #172.

erikwijmans commented on habitat_baselines/rl/ppo/ppo_trainer.py:
```diff
@@ -0,0 +1,519 @@
+                        ppo_cfg.lr,
+                    )
+
+                self.agent.clip_param = ppo_cfg.clip_param * (
```
This needs an `if ppo_cfg.use_linear_clip_decay` around it.

erikwijmans commented on habitat_baselines/rl/ppo/ppo_trainer.py:
```diff
@@ -0,0 +1,519 @@
+
+        ckpt = torch.load(checkpoint_path, map_location=device)
+
+        actor_critic = Policy(
```
Use `self._setup` here?

JasonJiazhiZhang replied:
Thanks for spotting that. I'll actually load saved config and merge them with config specified on evaluation.

erikwijmans commented on habitat_baselines/rl/ppo/ppo_trainer.py:
```diff
@@ -0,0 +1,519 @@
+
+
+@baseline_registry.register_trainer(name="ppo")
+class PPOTrainer(BaseRLTrainer):
```
Since the actually update rule (in this case PPO) is in a separate class, the vast majority of this trainer is really as `ActorCriticTrainer` and more general than just PPO.

JasonJiazhiZhang replied:
This definitely makes sense. I can make a separate PR for this structural change.

abhiskk commented on habitat_baselines/common/base_trainer.py:
```diff
@@ -0,0 +1,55 @@
+
+    supported_tasks: ClassVar[List[str]]
+
+    def __init__(self):
```
You don't need to write out the constructor explicitly.

abhiskk commented on habitat_baselines/common/env_utils.py:
```diff
@@ -0,0 +1,104 @@
+        env_class: class type of the envs to be created.
+
+    Returns:
+        vectorEnv object created according to specification.
```
nit: V should be capital

abhiskk commented on habitat_baselines/common/env_utils.py:
```diff
@@ -0,0 +1,104 @@
+            "reduce the number of processes as there "
+            "aren't enough number of scenes"
+        )
+        scene_split_size = int(
```
This variable is no longer needed

abhiskk commented on habitat_baselines/config/pointnav/ppo.yaml:
```diff
@@ -0,0 +1,40 @@
+      video_dir: "video_dir"
+      model_path: "data/checkpoints/ckpt.123.pth"
+      tracking_model_dir: ""
+      # 1 of the above 2 options needs to be empty
```
Is there an explicit check for this somewhere.

JasonJiazhiZhang replied:
I did a check [here](https://github.com/JasonJiazhiZhang/habitat-api/blob/c672710abfa3c58cce086ff623afb92596f7f47f/habitat_baselines/rl/ppo/ppo_trainer.py#L291-L293).

abhiskk commented on habitat_baselines/common/utils.py:
```diff
@@ -5,12 +5,60 @@
 # LICENSE file in the root directory of this source tree.
```
This file has become overloaded with a significant amount of code, distributing it even in a simple way across files will help in keeping the modularity. This can be done in this PR or can be added as a TODO for a later PR, @JasonJiazhiZhang  you can take a call on this.

JasonJiazhiZhang replied:
Let's do this in a later PR!

abhiskk commented on habitat_baselines/rl/ppo/ppo_trainer.py:
```diff
@@ -0,0 +1,519 @@
+import os
```
This is also another file which is really heavy in terms of code size, I have made suggestions below to abstract out the logics to other modules.

JasonJiazhiZhang replied:
Discussed offline. Will make more abstraction to `ppo_trainer.py`in a separate PR, stuff like moving ppo update and rollout stage out of `train()`

mathfac replied:
Agree with @abhiskk and @JasonJiazhiZhang need refactoring in next PR.

abhiskk commented on habitat_baselines/rl/ppo/ppo_trainer.py:
```diff
@@ -0,0 +1,519 @@
+        episode_spl_mean = aggregated_stats["spl"] / num_episodes
+        episode_success_mean = aggregated_stats["success"] / num_episodes
+
+        logger.info(
```
The logging and writer accesses below should be abstracted out to a separate util method.

abhiskk commented on habitat_baselines/rl/ppo/ppo_trainer.py:
```diff
@@ -0,0 +1,519 @@
+                    )
+
+    def _eval_checkpoint(self, checkpoint_path, args, writer, cur_ckpt_idx=0):
+        device = torch.device("cuda", args.pth_gpu_id)
```
This function needs to be broken down into separate modular methods.

abhiskk commented on habitat_baselines/rl/ppo/ppo_trainer.py:
```diff
@@ -0,0 +1,519 @@
+        return torch.load(checkpoint_path, map_location=self.device)
+
+    def train(self):
+        assert (
```
The train method for PPO is really heavy, we should abstract out logic present inside this class to separate methods.

abhiskk commented on habitat_baselines/rl/ppo/ppo_trainer.py:
```diff
@@ -0,0 +1,519 @@
+        ), "trainer is not properly initialized, need to specify config file"
+
+        ppo_cfg = self.config.TRAINER.RL.PPO
+        logger.info(
```
We should abstract out the logging logic to a single function call, having it strewn across the train method makes things messy.

abhiskk commented on habitat_baselines/rl/ppo/ppo_trainer.py:
```diff
@@ -0,0 +1,519 @@
+                    1 - update / ppo_cfg.num_updates
+                )
+
+                for step in range(ppo_cfg.num_steps):
```
The stepping logic below should be abstracted out across function calls.

JasonJiazhiZhang commented:
Updated PR: fixing merge conflict and modified README.
Changes from @abhiskk 's suggestion:
  * RolloutStorage now is in it's own file

Changes from @mathfac 's comments:
  * Now trainer config won't be passed into envs as a child config node.
  * Saved checkpoint config will be loaded to provide more info at evaluation time. 
  * Merging of saved config and eval config now has proper order: `eval_opts > saved_opts > eval_base_config > saved_base_config`.

mathfac commented on habitat_baselines/common/env_utils.py:
```diff
@@ -0,0 +1,104 @@
+    """
+    trainer_config = config.TRAINER.RL.PPO
+    rl_env_config = config.TRAINER.RL
+    config = config.TASK_CONFIG  # exclude trainer-specific configs
```
Usually one variable is used for one meaning, re-assign config isn't good idea, as it can have totally different meaning (experiment/baseline config and task config.) depends on code line.

mathfac commented on habitat_baselines/common/env_utils.py:
```diff
@@ -0,0 +1,101 @@
+    return env
+
+
+def construct_envs(config: Config, env_class: Type) -> VectorEnv:
```
Above comment is still relevant.

mathfac commented on habitat_baselines/config/default.py:
```diff
@@ -17,48 +17,87 @@
 _C = CN()
-_C.SEED = 100
+_C.BASE_TASK_CONFIG_PATH = "configs/tasks/pointnav.yaml"
+_C.TASK_CONFIG = get_config()  # task_config will be stored as a config node
```
Task config will be overwritten with default config otherwise:
```suggestion
_C.TASK_CONFIG = CN()  # task_config will be stored as a config node
```

JasonJiazhiZhang replied:
I agree, this makes the behavior more obvious for user. But this does prohibit user from overwrite task_config in `ppo.yaml`

mathfac commented on habitat_baselines/rl/ppo/ppo_trainer.py:
```diff
@@ -0,0 +1,568 @@
+import os
```
Add Facebook headers.

mathfac commented on habitat_baselines/run.py:
```diff
@@ -0,0 +1,67 @@
+    config = Config(new_allowed=True)
+    config.merge_from_other_cfg(baseline_cfg(cfg_path))
+    task_config = get_config(config.BASE_TASK_CONFIG_PATH)
+    config.TASK_CONFIG = task_config
```
You kill `config.TASK_CONFIG` node that was defined before. I think we discussed opposite:
`config.TASK_CONFIG.merge(task_config)`.

JasonJiazhiZhang replied:
Maybe `config.TASK_CONFIG` should take more priority in merging than base `task config`? Like:
```
config.TASK_CONFIG = task_config.merge(config.TASK_CONFIG)
```

mathfac commented on habitat_baselines/run.py:
```diff
@@ -0,0 +1,67 @@
+from habitat_baselines.config.default import get_config as baseline_cfg
+
+
+def get_exp_config(cfg_path: str, opts: List[str] = None) -> Config:
```
Why not to replace `get_config` for baselines with this function?
